src/shared/sam/cli/samCliInvoker.ts

```diff
@@ -8,42 +8,168 @@
+import {
+    DefaultSamCliUtils,
+    SamCliInfoResponse,
+    SamCliProcessInfo,
+    SamCliProcessInvoker,
+    SamCliTaskInvoker,
+    SamCliUtils } from './samCliInvokerUtils'
 import { DefaultSamCliLocationProvider } from './samCliLocator'
+import {
+    DefaultSamCliVersionValidator,
+    SamCliVersionValidation,
+    SamCliVersionValidator,
+    SamCliVersionValidatorResult
+} from './samCliVersionValidator'
 
-export interface SamCliProcessInvoker {
-    invoke(options: SpawnOptions, ...args: string[]): Promise<ChildProcessResult>
-    invoke(...args: string[]): Promise<ChildProcessResult>
+interface SamCliProcessInvokerContext {
+    cliConfig: SamCliConfiguration
+    cliInfo: SamCliProcessInfo
+    cliUtils: SamCliUtils
+    validator: SamCliVersionValidator
 }
 
-export interface SamCliTaskInvoker {
-    invoke(task: vscode.Task): Promise<vscode.TaskExecution>
+export function makeSamCliProcessInvokerContext(
+    params?: Partial<SamCliProcessInvokerContext>
+): SamCliProcessInvokerContext {
+    if (!params) {
+        params = {}
+    }
+
+    return {
+        cliConfig: params.cliConfig || new DefaultSamCliConfiguration(
+            new DefaultSettingsConfiguration(extensionSettingsPrefix),
+            new DefaultSamCliLocationProvider()
+        ),
+        cliInfo: params.cliInfo ||  { info: undefined, lastModified: undefined },
+        cliUtils: params.cliUtils || new DefaultSamCliUtils(),
+        validator: params.validator || new DefaultSamCliVersionValidator(),
+    }
 }
 
 export class DefaultSamCliProcessInvoker implements SamCliProcessInvoker {
-    public constructor(private readonly config: SamCliConfiguration = new DefaultSamCliConfiguration(
-        new DefaultSettingsConfiguration(extensionSettingsPrefix),
-        new DefaultSamCliLocationProvider()
-    )) {
-    }
+
+    public constructor(private readonly _context: SamCliProcessInvokerContext) {}
 
     public invoke(options: SpawnOptions, ...args: string[]): Promise<ChildProcessResult>
     public invoke(...args: string[]): Promise<ChildProcessResult>
     public async invoke(first: SpawnOptions | string, ...rest: string[]): Promise<ChildProcessResult> {
-        const args = typeof first === 'string' ? [ first, ...rest ] : rest
-        const options: SpawnOptions | undefined = typeof first === 'string' ? undefined : first
+        const logger: Logger = getLogger()
 
-        const samCliLocation = this.config.getSamCliLocation()
+        const samCliLocation = this._context.cliConfig.getSamCliLocation()
         if (!samCliLocation) {
-            throw new Error('SAM CLI location not configured')
+            const err = new Error('SAM CLI location not configured')
+            logger.error(err)
+            throw err
+        }
+        const validationResult = await this.getCliValidation(samCliLocation)
+
+        if (validationResult.validation === SamCliVersionValidation.Valid) {
```

Comment:
IMO this logic belongs in the validator. The invoker should just call `validator.validate`, catch any error that it throws, and notify the user.

Reply:
The problem is the `this.getCliValidation` command, since that would make another dependency loop.